### PR TITLE
Implement DataFrame.insert

### PR DESCRIPTION
**Pull request by xinrong-databricks**
ref #1929

Insert column into DataFrame at a specified location.

```
        >>> kdf = ks.DataFrame([1, 2, 3])
        >>> kdf.insert(0, 'x', 4)
        >>> kdf.sort_index()
           x  0
        0  4  1
        1  4  2
        2  4  3

        >>> from databricks.koalas.config import set_option, reset_option
        >>> set_option("compute.ops_on_diff_frames", True)

        >>> kdf.insert(1, 'y', [5, 6, 7])
        >>> kdf.sort_index()
           x  y  0
        0  4  5  1
        1  4  6  2
        2  4  7  3

        >>> kdf.insert(2, 'z', ks.Series([8, 9, 10]))
        >>> kdf.sort_index()
           x  y   z  0
        0  4  5   8  1
        1  4  6   9  2
        2  4  7  10  3

        >>> reset_option("compute.ops_on_diff_frames")
```

### CODE REVIEW

File: databricks/koalas/frame.py [hunk 1]
```diff
@@ -3698,7 +3700,121 @@ def notnull(self) -> "DataFrame":
...
+
+        Examples
+        --------
+        >>> kdf = ks.DataFrame([1, 2, 3])
+        >>> kdf
+           0
+        0  1
+        1  2
+        2  3
+        >>> set_option("compute.ops_on_diff_frames", False)
```
Comment (itholic) [inline databricks/koalas/frame.py:3733]: Seems `set_option` is not imported somewhere above.

**Review by itholic — COMMENTED** [review]
I'll take a more closer look after passing the tests :)

File: databricks/koalas/frame.py (hunk 1 above)
Comment (xinrong-databricks) [reply databricks/koalas/frame.py]: Good catch! Thanks.

File: databricks/koalas/frame.py [hunk 2]
```diff
@@ -3698,7 +3700,121 @@ def notnull(self) -> "DataFrame":
...
+           0
+        0  1
+        1  2
+        2  3
+        >>> kdf.insert(0, 'x', 4)
+        >>> kdf
+           x  0
+        0  4  1
+        1  4  2
+        2  4  3
```
[the remark this answers is not in the document]
Comment (xinrong-databricks) [reply databricks/koalas/frame.py:3738]: Certainly!

File: databricks/koalas/tests/test_ops_on_diff_frames.py [hunk 3]
```diff
@@ -477,6 +477,23 @@ def test_combine_first(self):
             kser1.combine_first(kser2).sort_index(), pser1.combine_first(pser2).sort_index()
         )
 
+    def test_insert(self):
+        pdf = pd.DataFrame([1, 2, 3])
+        kdf = ks.from_pandas(pdf)
+
+        # Inserta a Series (with a different anchor)
```
[the remark this answers is not in the document]
Comment (xinrong-databricks) [reply databricks/koalas/tests/test_ops_on_diff_frames.py]: Modified!

File: databricks/koalas/tests/test_dataframe.py [hunk 4]
```diff
@@ -88,6 +88,34 @@ def test_dataframe(self):
...
         self.assert_eq(kdf[index_cols], pdf[index_cols])
 
+    def test_insert(self):
+        pdf = pd.DataFrame([1, 2, 3])
+        kdf = ks.from_pandas(pdf)
+
+        # Inserts a Scalar
+        kdf.insert(1, "b", 10)
+        pdf.insert(1, "b", 10)
+        self.assert_eq(kdf.sort_index(), pdf.sort_index(), almost=True)
```
Comment (ueshin) [inline databricks/koalas/tests/test_dataframe.py:98]: Why `almost=True`?

File: databricks/koalas/frame.py [hunk 5]
```diff
@@ -3698,7 +3700,118 @@ def notnull(self) -> "DataFrame":
...
+
+        if not is_name_like_value(column):
+            raise ValueError(
+                "'column' should be a scalar value or tuple that contains scalar values"
+            )
+
+        if column in self.columns:
+            raise ValueError("cannot insert %s, already exists" % column)
+
+        is_value_scalar = isinstance(value, Scalar.__args__)  # type: ignore
```
Comment (ueshin) [inline databricks/koalas/frame.py:3773]: `not is_list_like(value)` or something instead of `isinstance(value, Scalar.__args__)`?

**Review by ueshin — COMMENTED** [review]
I found some bugs:

```py
>>> kdf = ks.DataFrame([1, 2, 3], index=[10,20,30])
>>> kdf.insert(1, 'z', ks.Series([8, 9, 10]))
>>> kdf
      0     z
0   NaN   8.0
1   NaN   9.0
10  1.0   NaN
2   NaN  10.0
30  3.0   NaN
20  2.0   NaN

>>> pdf = pd.DataFrame([1, 2, 3], index=[10,20,30])
>>> pdf.insert(1, 'z', pd.Series([8, 9, 10]))
>>> pdf
    0   z
10  1 NaN
20  2 NaN
30  3 NaN
```

or

```py
>>> kdf = ks.DataFrame({('x', 'a'): [1, 2, 3]})
>>> kdf.insert(1, 'z', ks.Series([8, 9, 10]))
Traceback (most recent call last):
...
AssertionError: [('x', 'a'), ('z',)]

>>> pdf = pd.DataFrame({('x', 'a'): [1, 2, 3]})
>>> pdf.insert(1, 'z', pd.Series([8, 9, 10]))
>>> pdf
   x   z
   a
0  1   8
1  2   9
2  3  10
```

File: databricks/koalas/tests/test_dataframe.py [hunk 6]
```diff
@@ -88,6 +88,34 @@ def test_dataframe(self):
...
+
+        kser = ks.Series([4, 5, 6])
+        self.assertRaises(ValueError, lambda: kdf.insert(0, "y", kser))
+        self.assertRaises(ValueError, lambda: kdf.insert(1, "y", kser))
+        self.assertRaises(ValueError, lambda: kdf.insert(0, list("abc"), kser))
+        self.assertRaises(ValueError, lambda: kdf.insert(0, "d", [7, 8, 9, 10]))
+        self.assertRaises(ValueError, lambda: kdf.insert(0, "d", ks.Series([7, 8])))
+
+        self.assertRaises(AssertionError, lambda: kdf.insert(100, "y", kser))
+        self.assertRaises(AssertionError, lambda: kdf.insert(1, "y", kser, allow_duplicates=True))
```
Comment (ueshin) [inline databricks/koalas/tests/test_dataframe.py:117]: Could you add some more tests using multi-index columns?

File: databricks/koalas/tests/test_ops_on_diff_frames.py [hunk 7]
```diff
@@ -477,6 +477,53 @@ def test_combine_first(self):
...
+
+        pser = pd.Series([4, 5, 6])
+        kser = ks.from_pandas(pser)
+        kdf.insert(1, ("y", ""), kser)
+        pdf.insert(1, ("y", ""), pser)
+        self.assert_eq(kdf.sort_index(), pdf.sort_index())
+
+        kdf.insert(1, "z", kser)
+        pdf.insert(1, "z", pser)
+        self.assert_eq(kdf.sort_index(), pdf.sort_index())
```
Comment (ueshin) [inline databricks/koalas/tests/test_ops_on_diff_frames.py:525]: ditto.

File: databricks/koalas/tests/test_dataframe.py (hunk 4 above)
Comment (xinrong-databricks) [reply databricks/koalas/tests/test_dataframe.py:98]: The inserted column dtype is different between Koalas and pandas.

After removing `almost=True`:
```
Attribute "dtype" are different
[left]:  int32
[right]: int64

Left:
   0   b
0  1  10
1  2  10
2  3  10
0    int64
b    int32
dtype: object

Right:
   0   b
0  1  10
1  2  10
2  3  10
0    int64
b    int64
dtype: object
```

File: databricks/koalas/tests/test_dataframe.py (hunk 6 above)
Comment (xinrong-databricks) [reply databricks/koalas/tests/test_dataframe.py:121]: Sounds good!

File: databricks/koalas/tests/test_ops_on_diff_frames.py (hunk 7 above)
Comment (xinrong-databricks) [reply databricks/koalas/tests/test_ops_on_diff_frames.py:520]: Thanks!

File: databricks/koalas/frame.py (hunk 5 above)
Comment (xinrong-databricks) [reply databricks/koalas/frame.py]: Thanks! I removed this line in the current implementation.

File: databricks/koalas/frame.py [hunk 8]
```diff
@@ -3698,7 +3699,88 @@ def notnull(self) -> "DataFrame":
...
+        if column in self.columns:
+            raise ValueError("cannot insert %s, already exists" % column)
+
+        kdf = self.copy()
+        kdf[column] = value
+        columns = kdf.columns[:-1].insert(loc, kdf.columns[-1])
+        kdf = kdf[columns]
+        self._update_internal_frame(kdf._internal)
+
+    # TODO: add frep and axis parqameter
```
Comment (ueshin) [inline databricks/koalas/frame.py:3783]: Shall we revert this change?

Comment (xinrong-databricks) [reply databricks/koalas/frame.py:3783]: May I ask why?

Comment (xinrong-databricks) [reply databricks/koalas/frame.py:3783]: I was thinking `kdf` and `self` should share the same anchor.

Comment (ueshin) [reply databricks/koalas/frame.py:3783]: I mean this line `# TODO: add frep and axis parqameter`

Comment (ueshin) [reply databricks/koalas/frame.py:3783]: btw

> I was thinking kdf and self should share the same anchor.

`kdf` could have a different anchor if the `value` is not the same anchor.

File: databricks/koalas/frame.py [hunk 9]
```diff
@@ -3698,7 +3699,87 @@ def notnull(self) -> "DataFrame":
 
     notna = notnull
 
-    # TODO: add frep and axis parameter
```
Comment (ueshin) [inline databricks/koalas/frame.py:3701]: Could you add this line back?

File: databricks/koalas/frame.py (hunk 8 above)
Comment (xinrong-databricks) [reply databricks/koalas/frame.py]: I see, may I get some help understand `requires_same_anchor` parameter of [`_update_internal_frame`](https://github.com/databricks/koalas/blob/master/databricks/koalas/frame.py#L525) ?

File: databricks/koalas/frame.py (hunk 9 above)
Comment (xinrong-databricks) [reply databricks/koalas/frame.py:3701]: Added!

File: databricks/koalas/frame.py (hunk 8 above)
Comment (ueshin) [reply databricks/koalas/frame.py]: If it's `False`, the existing linked `Series` will see the updates even when the anchor changes; otherwise the link will be disconnected.
See #1592 for more detail.

Comment (xinrong-databricks) [reply databricks/koalas/frame.py]: Thanks for the reference!

No matter what `requires_same_anchor` is, the test case doesn't seem to pass https://github.com/databricks/koalas/pull/1983/files#diff-028bf26a42786beb47e6707fe34867dc720d3279ae4c942226abc3eb40f26eeaR104.
Would you give me a clue?

File: databricks/koalas/tests/test_dataframe.py [hunk 10]
```diff
@@ -88,6 +88,68 @@ def test_dataframe(self):
...
+        pdf = pd.DataFrame([1, 2, 3])
+        kdf = ks.from_pandas(pdf)
+
+        kdf.insert(1, "b", 10)
+        pdf.insert(1, "b", 10)
+        self.assert_eq(kdf.sort_index(), pdf.sort_index(), almost=True)
+
+        kdf.b = kdf.b + 1
+        pdf.b = pdf.b + 1
+        self.assert_eq(kdf.sort_index(), pdf.sort_index(), almost=True)
```
Comment (xinrong-databricks) [inline databricks/koalas/tests/test_dataframe.py:104]: Here `kdf` doesn't recognize `kdf.b`'s change.

Comment (ueshin) [reply databricks/koalas/tests/test_dataframe.py:104]: I'm not sure what you mean? Seems like the tests passed.

Comment (ueshin) [reply databricks/koalas/tests/test_dataframe.py:104]: Btw, I'm also not sure what you want to test here.

Comment (xinrong-databricks) [reply databricks/koalas/tests/test_dataframe.py:104]: In my local,
```
Left: # koalas
   0   b
0  1  10
1  2  10
2  3  10
0    int64
b    int32
dtype: object

Right: # pandas
   0   b
0  1  11
1  2  11
2  3  11
0    int64
b    int64
dtype: object
```
After column b is inserted to kdf, `kdf.b = kdf.b + 1` can't be recognized by `kdf`, so column `b` remains 10 rather than 11 in the example.

Comment (xinrong-databricks) [reply databricks/koalas/tests/test_dataframe.py:104]: It surprised me CI jobs pass now.

Comment (xinrong-databricks) [reply databricks/koalas/tests/test_dataframe.py:104]:
```
>>> kdf = ks.DataFrame({'x': [1, 2, 3]})
>>> kdf
   x
0  1
1  2
2  3
>>> kdf.x = kdf.x + 1
>>> kdf = ks.DataFrame({'x': [1, 2, 3]})
>>> kdf.x = kdf.x + 1
>>> kdf.x
0    2
1    3
2    4
Name: x, dtype: int64
>>> kdf # wondering why
   x
0  1
1  2
2  3
>>> pdf = pd.DataFrame({'x': [1, 2, 3]})
>>> pdf.x = pdf.x + 1
>>> pdf.x
0    2
1    3
2    4
Name: x, dtype: int64
>>> pdf 
   x
0  2
1  3
2  4
```

Comment (ueshin) [reply databricks/koalas/tests/test_dataframe.py:104]: I guess you should merge master? Seems like this branch doesn't include the commit by #1989.

Comment (xinrong-databricks) [reply databricks/koalas/tests/test_dataframe.py:104]: Got it, thanks!

Comment (ueshin) [inline databricks/koalas/tests/test_dataframe.py:104]: I guess we don't need this test.

File: databricks/koalas/tests/test_ops_on_diff_frames.py [hunk 11]
```diff
@@ -477,6 +477,52 @@ def test_combine_first(self):
...
+
+        pser = pd.Series([4, 5, 6])
+        kser = ks.from_pandas(pser)
+        kdf.insert(1, "y", kser)
+        pdf.insert(1, "y", pser)
+        self.assert_eq(kdf.sort_index(), pdf.sort_index())
+
+        kdf.y = kdf.y + 1
+        pdf.y = pdf.y + 1
+        self.assert_eq(kdf.sort_index(), pdf.sort_index())
```
Comment (ueshin) [inline databricks/koalas/tests/test_ops_on_diff_frames.py:495]: ditto.

**Review by ueshin — APPROVED** [review]
Otherwise, LGTM.

**Comment by ueshin** [comment]
Thanks! merging.

**Comment by xinrong-databricks** [comment]
@ueshin Thank you!
